diff -Naurp old/kexec/arch/i386/kexec-multiboot-x86.c 
new/kexec/arch/i386/kexec-multiboot-x86.c
--- old/kexec/arch/i386/kexec-multiboot-x86.c   2007-02-27 11:00:34.000000000 +0530
+++ new/kexec/arch/i386/kexec-multiboot-x86.c   2007-02-27 11:01:58.000000000 +0530
@@ -47,7 +47,6 @@
 #include <getopt.h>
 #include <elf.h>
 #include <boot/elf_boot.h>
-#include <asm/page.h>
 #include <ip_checksum.h>
 #include "../../kexec.h"
 #include "../../kexec-elf.h"
@@ -341,7 +340,7 @@ int multiboot_x86_load(int argc, char **
 			/* Pick the next aligned spot to load it in */
 			freespace = add_buffer(info,
 				buf, mod_size, mod_size,
-				PAGE_SIZE, 0, 0xffffffffUL, 1);
+				getpagesize(), 0, 0xffffffffUL, 1);
 
 			/* Add the module command line */
 			sprintf(mod_clp, "%s", mod_command_line);
